### PR TITLE
feat: add suggestions to array-callback-return

### PR DESCRIPTION
**Pull request by Tanujkanti4441**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[x] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)
Added suggestions to the `array-callback-return` rule.
The fixer functions of PR #17282 are used to implement suggestions.

#### Is there anything you'd like reviewers to focus on?

Fixes- #17285

### CODE REVIEW

File: lib/rules/array-callback-return.js [hunk 1]
```diff
@@ -136,21 +136,93 @@ function getArrayMethodName(node) {
...
+
 //------------------------------------------------------------------------------
 // Rule Definition
 //------------------------------------------------------------------------------
 
 /** @type {import('../shared/types').Rule} */
 module.exports = {
     meta: {
-        type: "problem",
+        type: "suggestion",
```
Comment (mdjermanovic) [inline lib/rules/array-callback-return.js:216]:
```suggestion
        type: "problem",
```

This change should be reverted, as the `meta.type` field isn't related to providing editor suggestions. `type: "problem"` means that the rule reports possible errors in code, which is the case with this rule. `type: "suggestion"` means that the rule only enforces preferences. Both kinds of rules can have editor suggestions :)

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

**Comment by mdjermanovic** [comment]
Linting fails with this error:

```
Error:   224:25  error  `meta.hasSuggestions` cannot be `true` for non-suggestable rules  eslint-plugin/require-meta-has-suggestions
```

Looks like `eslint-plugin/require-meta-has-suggestions` can't detect that this rule does provide suggestions. I think the right solution, at the moment, is to disable `eslint-plugin/require-meta-has-suggestions`:

```diff
+ // eslint-disable-next-line eslint-plugin/require-meta-has-suggestions -- false positive
hasSuggestions: true,
```

**Comment by Tanujkanti4441** [comment]
> Linting fails with this error:
> 
> ```
> Error:   224:25  error  `meta.hasSuggestions` cannot be `true` for non-suggestable rules  eslint-plugin/require-meta-has-suggestions
> ```
> 
> Looks like `eslint-plugin/require-meta-has-suggestions` can't detect that this rule does provide suggestions. I think the right solution, at the moment, is to disable `eslint-plugin/require-meta-has-suggestions`:
> 
> ```diff
> + // eslint-disable-next-line eslint-plugin/require-meta-has-suggestions -- false positive
> hasSuggestions: true,
> ```

Wow this saves me. I am still wondering what causing this error even though all tests are clear.

File: lib/rules/array-callback-return.js [hunk 2]
```diff
@@ -209,32 +284,57 @@ module.exports = {
                 return;
             }
 
-            let messageId = null;
+            // let messageId = null;
```
Comment (mdjermanovic) [inline lib/rules/array-callback-return.js:287]:
```suggestion
```

We can remove this if it's not needed.

File: lib/rules/array-callback-return.js [hunk 3]
```diff
@@ -295,36 +395,46 @@ module.exports = {
...
                     context.report({
                         node,
-                        messageId,
+                        messageId: messageAndSuggestions.messageId,
                         data: {
                             name: astUtils.getFunctionNameWithKind(funcInfo.node),
                             arrayMethodName: fullMethodName(funcInfo.arrayMethodName)
-                        }
+                        },
+                        suggest: messageAndSuggestions.suggest !== 0 ? messageAndSuggestions.suggest : null
```
Comment (mdjermanovic) [inline lib/rules/array-callback-return.js:437]:
```suggestion
                        suggest: messageAndSuggestions.suggest.length !== 0 ? messageAndSuggestions.suggest : null
```

File: tests/lib/rules/array-callback-return.js [hunk 4]
```diff
@@ -239,6 +239,85 @@ ruleTester.run("array-callback-return", rule, {
...
         { code: "foo.forEach((x) => { if (a === b) { return !x } })", options: checkForEachAllowVoid, parserOptions: { ecmaVersion: 6 }, errors: [{ messageId: "expectedNoReturnValue" }] },
 
+        {
+            code: "foo.forEach(x => x)",
+            options: checkForEachOptions,
+            parserOptions: { ecmaVersion: 6 },
+            errors: [{
+                messageId: "expectedNoReturnValue",
+                suggestions: [
+                    { output: "foo.forEach(x => {x})" }
```
Comment (mdjermanovic) [inline tests/lib/rules/array-callback-return.js:249]: Can you add `messageId` to all suggestion test objects? Also, can you add `suggestions` to all test cases where the rule provides suggestions? This is for better testing and will soon be mandatory per [RFC #84](https://github.com/eslint/rfcs/blob/main/designs/2021-stricter-rule-test-validation/README.md).

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!

**Pull request merged**